### PR TITLE
Add functions to detect multi-track windows

### PR DESCRIPTION
**Pull request by pphuangyi**
I added a function that determines whether an image contains multiple tracks.
Please feel free to comment.

### CODE REVIEW

**Review by brettviren — APPROVED** [review]
I think this is fine, assuming your tests show it works the way you want it to.  I think it is fine as is but I offer some comments.

This is doing a form of "event reconstruction" which is very challenging in general as it has many corner cases where things do not work.

If I understand the code, I expect some corner cases where it will not work as expected.

- it is possible, though rare, that two tracks will overlap at their endpoints such that your `scan()` does not see more than one run of 1's.
- in realistic simulation we will have a lot of activity associated with a track that is separated in pixel-space from the "main" track.  Depending on how one interprets "multi" this associated activity should or should not be considered separate.

Also, just curious if `scan()` becomes a performance bottleneck?  Any time Python drops down to dealing with one index of an array at a time it tends to be slow.  If it is not a problem, we do not need to make it a problem.  But, if it is a problem perhaps "vectorizing" the function would help:

https://numpy.org/doc/stable/reference/generated/numpy.vectorize.html

**Comment by usert5432** [comment]
Hi Yi. Thank you very much for implementing the multitrack detector! Your method is very fast, simple, and elegant, and I think it should be sufficient for our purposes. I have a few suggestions about the implementation, that I will leave below.

File: toytools/transform.py [hunk 1]
```diff
@@ -72,3 +72,68 @@ def try_find_region_with_signal(
...
+    one = False
+    for v in vec:
+        if v == 0 and one:
+            count += 1
+            one = False
+            continue
+        if v == 1 and not one:
+            one = True
+            continue
+    return count + int(one)
```
Comment (usert5432) [inline toytools/transform.py]: Since you are dealing with images of 0s and 1s, you can detect tracks by presence of a rising edge. For example, here is a `vec` containing a single track `[0, 0, 1, 1, 1, 0, 0]`. Here, values change from 0 to 1 at the beginning of the track (going from index 1 to index 2). So, by counting the number of such changes you can find the number of tracks.

The following expression should give you the number of tracks:
`np.count_nonzero((v[1:] - v[:-1]) == 1)`

Additional benefit is that this expression could be easily broadcasted over the entire image, improving the detection speed.

File: toytools/transform.py [hunk 2]
```diff
@@ -72,3 +72,68 @@ def try_find_region_with_signal(
...
+		- kernel_size (int): the kernel size of torch.AvgPool2d
+		- stride (int): the stride of torch.AvgPool2d
+	Output (numpy.ndarray): the processed 0-1 image
+	"""
+    x = np.abs(x)
+    x[x > 0] = 1
+    with torch.no_grad():
+        x = torch.tensor(x).to(torch.float)
+        input = x.view(1, 1, x.shape[0], x.shape[1])
+        avgpool = nn.AvgPool2d(kernel_size, stride=stride)
```
Comment (usert5432) [inline toytools/transform.py]: Could you please remind me why you need pooling? Sorry, I remember you explained it during the meeting, but I cannot recall what you have said.

File: toytools/transform.py [hunk 3]
```diff
@@ -72,3 +72,68 @@ def try_find_region_with_signal(
...
+	Preprocessing steps for multitrack detections.
+	Input:
+		- x (numpy.ndarray): the input image of shape (H, W)
+		- kernel_size (int): the kernel size of torch.AvgPool2d
+		- stride (int): the stride of torch.AvgPool2d
+	Output (numpy.ndarray): the processed 0-1 image
+	"""
+    x = np.abs(x)
+    x[x > 0] = 1
+    with torch.no_grad():
```
Comment (usert5432) [inline toytools/transform.py]: I think you need to add `import torch` to the beginning of the file, otherwise it will complain about undefined variable `torch`.

Comment (usert5432) [reply toytools/transform.py]: Also, I think we have decided before not to use `torch` in data processing functions. We can easily revise that decision, but I was wondering if pooling can be implemented using the `numpy` only functions?

**Comment by usert5432** [comment]
@pphuangyi, could I ask you please also to run a `pylint` program over the file `toytools/transform.py`, like:
```
$ pylint toytools/transform.py
```

and make the changes that `pylint` suggests? Sorry, it is usually very tedious to fix its suggestions, but it will help us to keep the code style consistent across projects.

File: toytools/transform.py [hunk 4]
```diff
@@ -72,3 +72,68 @@ def try_find_region_with_signal(
...
+            count += 1
+            one = False
+            continue
+        if v == 1 and not one:
+            one = True
+            continue
+    return count + int(one)
+
+
+def isMultiTrack(x, kernel_size=2, stride=2, fraction=.05):
```
Comment (usert5432) [inline toytools/transform.py]: The name `isMultiTrack` uses camelCase. I do not think we have discussed that, but I believe we would want our functions to be named in snake_case. @brettviren, @YHRen do you agree about the function naming style?

Comment (brettviren) [reply toytools/transform.py]: Yes, `snake_case` for functions and variables is what I commonly see (and prefer) in Python.

**Comment by pphuangyi** [comment]
Hi Dmitrii and Brett,
Sorry for the late reply. I did a few tests on vectorization. It didn't bring speed up on my old version of the `scan`. But I find out that my old scan is doing much more than needed. Old `scan` finds the number of consecutive sections of 1s in a vector, but for determining whether a second track exists, we can just exit the scan as soon as two such sections are found. So  I pushed a new version: https://github.com/LS4GAN/toytools/blob/9c98cc912709a19a85d33eb0f0cdb5dd09bf4a9c/toytools/transform.py#L121
and it does bring quite a lot of speedup in _comfirming a multitrack window_. However, it is still not fast enough in disproving a window as having multiple tracks. I will have to think more about it.

One idea is to only sample a few columns and rows. I will experiment on them later today or tomorrow.

File: toytools/transform.py (hunk 4 above)
Comment (pphuangyi) [reply toytools/transform.py]: Sorry I missed this discussion. I will use `snake_case` instead and update again.

File: toytools/transform.py (hunk 3 above)
Comment (pphuangyi) [reply toytools/transform.py]: Thank you, Dmitrii! I temporarily add `torch` to the file. But I agree with you in that we shouldn't use `torch` here and will try a `torch`-free version.

File: toytools/transform.py (hunk 2 above)
Comment (pphuangyi) [reply toytools/transform.py]: Hi Dmitrii, occasionally (and rarely), even when there is a single track, there can be a zero streak in the middle of a track (probably since the data is bipolar), in which case the scan will think there are multi-tracks. But it is probably really really rare. I use pooling to average things out so that the zero will disappear. But I agree that it is probably not very necessary.

File: toytools/transform.py (hunk 1 above)
Comment (pphuangyi) [reply toytools/transform.py]: Indeed! I will adopt this suggestion!

Comment (pphuangyi) [reply toytools/transform.py]: Hi Dmitrii,

I did `pylint` and correct most of the suggestions. However, I don't know what to do with these last two:
`transform.py:92:12: E1101: Module 'torch' has no 'as_tensor' member; maybe 'is_tensor'? (no-member)`
`transform.py:92:34: E1101: Module 'torch' has no 'float32' member (no-member)`

I've tested that the code works fine. Do you have any idea?

Comment (pphuangyi) [reply toytools/transform.py]: Hi Dmitrii,

The speed almost doubled! Thank you so much! And the code is much simpler! 

I actually used `np.count_nonzero(np.abs(vec[1:] - vec[:-1])) >= 3` to cover the case of one track getting too close to the boundary.

https://github.com/LS4GAN/toytools/blob/aa8a116ca06509a4462ecabd63f7132f3f93a062/toytools/transform.py#L100-L108

File: toytools/transform.py (hunk 3 above)
Comment (pphuangyi) [reply toytools/transform.py]: Hi Dmitrii, I used a  `NumPy` version of average pooling. And it brought a **10x** speedup :) I am so excited! 

And `pylint` is also completely happy with the code (10/10).

File: toytools/transform.py (hunk 2 above)
Comment (brettviren) [reply toytools/transform.py]: This is a good point.  And, next sample when we add "real" Geant4 stepping through the LAr medium the tracks will reflect dE/dx fluctuations, multiple scattering, delta ray emissions, and more.  All these will cause pixels to dip below threshold when viewing just 1 view of the 2D image even though the "peaks" on either side of the "valley" might be considered connected.

IOW, `is_multitrack()` is a hard function to implement in general!

But, it is still practical to have a simple function even if "multitrack" is maybe open to different interpretations.

File: toytools/transform.py [hunk 5]
```diff
@@ -72,3 +72,66 @@ def try_find_region_with_signal(
...
+    x[x > 0] = 1
+    H, W = x.shape
+    H_k, W_k = H // kernel_size, W // kernel_size
+    new_shape = (H_k, kernel_size, W_k, kernel_size)
+    y = x[:H_k * kernel_size, :W_k * kernel_size].reshape(*new_shape).mean(axis=(1, 3))
+    y[y > 0] = 1
+    return y
+
+
+def scan(vec):
```
Comment (YHRen) [inline toytools/transform.py]: `scan` is a very general name, maybe move it inside `is_multi_track`.

**Review by YHRen — APPROVED** [review]
Very nice!

File: toytools/transform.py (hunk 1 above)
Comment (usert5432) [reply toytools/transform.py]: Thank you, Yi!

> I did pylint and correct most of the suggestions. However, I don't know what to do with these last two:
transform.py:92:12: E1101: Module 'torch' has no 'as_tensor' member; maybe 'is_tensor'? (no-member)
transform.py:92:34: E1101: Module 'torch' has no 'float32' member (no-member)

Fo reference, to use pylint with `torch` one needs to modify `pylint` config and add lines like

https://github.com/LS4GAN/toygan/blob/dff5079be698f439e3693278eb7beb2f9f16178b/.pylintrc#L16-L18

**Review by usert5432 — APPROVED** [review]

File: toytools/transform.py (hunk 2 above)
Comment (usert5432) [reply toytools/transform.py]: Thank you for the explanation, Yi!

File: toytools/transform.py (hunk 3 above)
Comment (usert5432) [reply toytools/transform.py]: Awesome, Thanks!
